ethpm_cli/scraper.py

```diff
@@ -0,0 +1,215 @@
+import itertools
+import json
+import logging
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Set, Tuple  # noqa: F401
+
+from eth_utils import to_dict, to_list
+from eth_utils.toolz import assoc
+from ethpm.typing import URI, Address
+from ethpm.utils.backend import resolve_uri_contents
+from ethpm.utils.ipfs import extract_ipfs_path_from_uri, is_ipfs_uri
+from ethpm.utils.uri import is_supported_content_addressed_uri
+from web3 import Web3
+
+from ethpm_cli._utils.various import flatten
+from ethpm_cli.constants import VERSION_RELEASE_ABI
+from ethpm_cli.exceptions import BlockAlreadyScrapedError, BlockNotFoundError
+from ethpm_cli.validation import validate_chain_data_store
+
+logger = logging.getLogger("ethpm_cli.scraper.Scraper")
+
+
+def scrape(w3: Web3, ethpm_dir: Path, start_block: int = 0) -> None:
+    """
+    Scrapes VersionRelease event data starting from start_block.
+    """
+    chain_data_path = ethpm_dir / "chain_data.json"
+    try:
+        active_block = import_ethpm_dir(ethpm_dir, w3, start_block)
+        scraped_manifests = scrape_block_for_manifests(w3, active_block)
+
+        update_chain_data(chain_data_path, active_block, scraped_manifests)
+        write_ipfs_uris_to_disk(ethpm_dir, scraped_manifests)
+
+    except BlockNotFoundError:
+        logger.info("No more blocks available.")
+        return None
+
+    except BlockAlreadyScrapedError:
+        active_block = start_block + 1
+
+    scrape(w3, ethpm_dir, active_block)
+
+
+def import_ethpm_dir(ethpm_dir: Path, w3: Web3, start_block: int) -> int:
+    """
+    Returns the latest processed block number found in ethpm_dir.
+    If no ethpm_dir found, creates an ethpm_dir and returns a 0.
+    """
+    if ethpm_dir.is_dir():
+        chain_data_path = ethpm_dir / "chain_data.json"
+        validate_chain_data_store(chain_data_path, w3)
+
+        if start_block >= w3.eth.blockNumber:
+            raise BlockNotFoundError(
+                f"Block number: {start_block} not available on provided web3 "
+                f"instance with latest block number of {w3.eth.blockNumber}."
```

Comment:
race condition here,  probably should pull off `w3.eth.blockNumber` to a local variable so that you don't encounter a situation where it changes between the two locations this property is accessed.